File: scripts/clean-data-gender.ts
import path from "path";

// Note: this script uses the Node polars bindings. Install with:
//    npm install polars
// or
//    yarn add polars
import pl from "nodejs-polars";

async function main() {
  const input = path.join(process.cwd(), "data", "sex", "collection.csv");
  const output = path.join(process.cwd(), "src", "data", "gender.csv");

  // Read CSV
  const df = pl.readCSV(input, { tryParseDates: false });

  // 1) Clean Filename: remove prefix/suffix
  const df1 = df.withColumns([
    pl
      .col("Filename")
      .str()
      .replaceAll("data/sex/", "")
      .replaceAll(".csv", "")
      .alias("Filename_clean"),
  ]);

  // 2) Split Filename_clean into Start Date and End Date, drop original
  const df2 = df1.withColumns([
    pl.col("Filename_clean").str().splitExact("-to-", 2).arr().get(0).alias("Start Date"),
    pl.col("Filename_clean").str().splitExact("-to-", 2).arr().get(1).alias("End Date"),
  ]).drop(["Filename", "Filename_clean"]);

  // 3) Reformat Start/End dates from YYYY-MM-DD to DD/MM/YYYY if needed
  const df3 = df2.withColumns([
    pl
      .col("End Date")
      .str()
      .replaceAll(/(\d{4})-(\d{2})-(\d{2})/, "$3/$2/$1")
      .alias("End Date"),
    pl
      .col("Start Date")
      .str()
      .replaceAll(/(\d{4})-(\d{2})-(\d{2})/, "$3/$2/$1")
      .alias("Start Date"),
  ]);

  // 4) One-hot encode 'Value'
  // polars JS exposes getDummies on a DataFrame
  // create dummies df for the Value column
  const dummies = df3.getDummies({ columns: ["Value"] });

  // Merge dummies back with the rest of the data (drop original Value then concat)
  const base = df3.drop(["Value"]);
  const merged = pl.concat([base, dummies], { how: "horizontal" });

  // 5) Rename awkward column if present
  let df4 = merged;
  if (df4.columns.includes("Value_would rather not say")) {
    df4 = df4.rename({
      "Value_would rather not say": "Value_would_rather_not_say",
    });
  }

  // 6) Ensure Count is numeric
  df4 = df4.withColumn(pl.col("Count").cast(pl.Int64));

  // 7) Convert boolean-like dummy columns to integers if necessary
  const boolToIntCols = df4.columns.filter((c) => c.startsWith("Value_"));
  const castExprs = boolToIntCols.map((c) =>
    pl.col(c).cast(pl.Int64).alias(c)
  );
  df4 = df4.withColumns(castExprs);

  // 8) Create aggregated gender columns
  const withGender = df4.withColumns([
    pl
      .col("Value_male")
      .fillNull(0)
      .mul(pl.col("Count"))
      .cast(pl.Int64)
      .alias("male"),
    pl
      .col("Value_female")
      .fillNull(0)
      .mul(pl.col("Count"))
      .cast(pl.Int64)
      .alias("female"),
    pl
      .col("Value_[None]")
      .fillNull(0)
      .mul(pl.col("Count"))
      .cast(pl.Int64)
      .alias("none"),
    pl
      .col("Value_other")
      .fillNull(0)
      .mul(pl.col("Count"))
      .cast(pl.Int64)
      .alias("other"),
    pl
      .col("Value_would_rather_not_say")
      .fillNull(0)
      .mul(pl.col("Count"))
      .cast(pl.Int64)
      .alias("would_rather_not_say"),
  ]);

  // 9) Drop intermediate Value_* and Amount
  const dropCols = [
    "Value_[None]",
    "Value_female",
    "Value_male",
    "Value_other",
    "Value_would_rather_not_say",
    "Amount",
  ].filter((c) => withGender.columns.includes(c));
  const df5 = withGender.drop(dropCols);

  // 10) Clone Start Date as start_date_time and convert format to YYYY-MM-DD
  const df6 = df5.withColumns([
    pl
      .col("Start Date")
      .alias("start_date_time")
      .str()
      .replaceAll(/^(\d{2})\/(\d{2})\/(\d{4})$/, "$3-$2-$1"),
  ]);

  // 11) Ensure numeric types
  const final = df6.withColumns([
    pl.col("female").cast(pl.Int64),
    pl.col("none").cast(pl.Int64),
    pl.col("male").cast(pl.Int64),
    pl.col("other").cast(pl.Int64),
    pl.col("would_rather_not_say").cast(pl.Int64),
  ]);

  // 12) Group by start_date_time and aggregate
  const grouped = final
    .groupBy(["start_date_time"])
    .agg([
      pl.col("Start Date").first().alias("StartDate_first"),
      pl.col("End Date").first().alias("EndDate_first"),
      pl.col("male").sum().alias("male_sum"),
      pl.col("female").sum().alias("female_sum"),
      pl.col("none").sum().alias("none_sum"),
      pl.col("would_rather_not_say").sum().alias("would_rather_not_say_sum"),
      pl.col("Count").sum().alias("Count_sum"),
    ])
    .sort("start_date_time");

  // Write out CSV
  await pl.writeCSV(grouped, output);

  console.log(`Wrote cleaned gender CSV to ${output}`);
}

if (require.main === module) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

File: scripts/clean_data_gender.py
import pandas as pd


def clean_data(df):
    # Clean date
    df["Filename"] = df["Filename"].str.replace(
        "data/sex/|\\.csv", "", case=False, regex=True
    )
    # Split text using string '-to-' in column: 'Filename'
    loc_0 = df.columns.get_loc("Filename")
    df_split = df["Filename"].str.split(pat="-to-", expand=True).add_prefix("Filename_")
    df = pd.concat([df.iloc[:, :loc_0], df_split, df.iloc[:, loc_0:]], axis=1)
    df = df.drop(columns=["Filename"])
    # Rename column 'Filename_0' to 'Start Date'
    df = df.rename(columns={"Filename_0": "Start Date"})
    # Rename column 'Filename_1' to 'End Date'
    df = df.rename(columns={"Filename_1": "End Date"})
    # Replace all instances of "(\\d\\d\\d\\d)-(\\d\\d)-(\\d\\d)" with "\\3/\\2/\\1" in columns: 'End Date', 'Start Date'
    df["End Date"] = df["End Date"].str.replace(
        "(\\d\\d\\d\\d)-(\\d\\d)-(\\d\\d)", "\\3/\\2/\\1", case=False, regex=True
    )
    df["Start Date"] = df["Start Date"].str.replace(
        "(\\d\\d\\d\\d)-(\\d\\d)-(\\d\\d)", "\\3/\\2/\\1", case=False, regex=True
    )
    # One-hot encode column: 'Value'
    insert_loc = df.columns.get_loc("Value")
    df = pd.concat(
        [
            df.iloc[:, :insert_loc],
            pd.get_dummies(df.loc[:, ["Value"]]),
            df.iloc[:, insert_loc + 1 :],
        ],
        axis=1,
    )

    # Replace all booleans with 0 or 1
    df = df.rename(columns={"Filename_0": "Start Date"})
    df = df.rename(columns={"Value_OTHER": "Value_other"})
    df = df.rename(columns={"Value_FEMALE": "Value_female"})
    df = df.rename(columns={"Value_MALE": "Value_male"})
    df = df.rename(columns={"Value_UNSPECIFIED": "Value_unspecified"})

    df.loc[df["Value_female"] == False, "Value_female"] = 0
    df.loc[df["Value_female"] == True, "Value_female"] = 1
    df.loc[df["Value_[None]"] == False, "Value_[None]"] = 0
    df.loc[df["Value_[None]"] == True, "Value_[None]"] = 1
    df.loc[df["Value_male"] == False, "Value_male"] = 0
    df.loc[df["Value_male"] == True, "Value_male"] = 1
    df.loc[df["Value_other"] == False, "Value_other"] = 0
    df.loc[df["Value_other"] == True, "Value_other"] = 1
    df.loc[df["Value_unspecified"] == False, "Value_unspecified"] = 0
    df.loc[df["Value_unspecified"] == True, "Value_unspecified"] = 1
    # create new columns
    df["male"] = df["Value_male"] * df["Count"]
    df["female"] = df["Value_female"] * df["Count"]
    df["none"] = df["Value_[None]"] * df["Count"]
    df["other"] = df["Value_other"] * df["Count"]
    df["unspecified"] = df["Value_unspecified"] * df["Count"]
    # Drop columns: 'Value_[None]', 'Value_female' and 3 other columns
    df = df.drop(
        columns=[
            "Value_[None]",
            "Value_female",
            "Value_male",
            "Value_other",
            "Value_unspecified",
        ]
    )
    # Drop column: 'Amount'
    df = df.drop(columns=["Amount"])
    # Clone column 'Start Date' as 'start_date_time'
    df["start_date_time"] = df.loc[:, "Start Date"]
    # Replace all instances of "(\\d\\d)/(\\d\\d)/(\\d\\d\\d\\d)" with "\\3-\\2-\\1" in column: 'start_date_time'
    df["start_date_time"] = df["start_date_time"].str.replace(
        "^(\\d\\d)/(\\d\\d)/(\\d\\d\\d\\d)$", "\\3-\\2-\\1", case=False, regex=True
    )
    # Change column type to int64 for column: 'female'
    df = df.astype({"female": "int64"})
    # Change column type to int64 for column: 'none'
    df = df.astype({"none": "int64"})
    # Change column type to int64 for column: 'male'
    df = df.astype({"male": "int64"})
    # Change column type to int64 for columns: 'other', 'unspecified'
    df = df.astype({"other": "int64", "unspecified": "int64"})
    # Performed 7 aggregations grouped on column: 'start_date_time'
    df = (
        df.groupby(["start_date_time"])
        .agg(
            StartDate_first=("Start Date", "first"),
            EndDate_first=("End Date", "first"),
            male_sum=("male", "sum"),
            female_sum=("female", "sum"),
            none_sum=("none", "sum"),
            unspecified_sum=("unspecified", "sum"),
            Count_sum=("Count", "sum"),
        )
        .reset_index()
    )
    return df


# Loaded variable 'df' from URI: /home/damien/projects/ltl-usage-graph/data/collection.csv
df = pd.read_csv(r"data/sex/collection.csv", engine="pyarrow")

df_clean = clean_data(df.copy())
df_clean.to_csv(r"src/data/gender.csv", index=False)
